### PR TITLE
[Merged by Bors] - feat(Combinatorics/SimpleGraph): Add lemma about number of odd subcomponents

### PR DESCRIPTION
**Pull request by pimotte**
Added lemma that the number of odd components contained within a component
in a bigger graph is odd iff the bigger component is odd. In preparation for Tuttes.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

[Zulip thread](https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/Tutte's.20theorem)

### CODE REVIEW

File: Mathlib/Combinatorics/SimpleGraph/Connectivity/WalkCounting.lean
Comment (YaelDillies) [inline Mathlib/Combinatorics/SimpleGraph/Connectivity/WalkCounting.lean:190]: Any reason why you are nesting set comprehensions here?
```suggestion
    {c : ConnectedComponent G | c.supp ⊆ c'.supp ∧ Odd (Nat.card c.1.supp)}) := by
```

File: Mathlib/Combinatorics/SimpleGraph/Path.lean
Comment (YaelDillies) [inline Mathlib/Combinatorics/SimpleGraph/Path.lean:972]: Maybe
```suggestion
lemma biUnion_supp_eq_supp {G G' : SimpleGraph V} (h : G ≤ G') (c' : ConnectedComponent G') :
    ⋃ (c : ConnectedComponent G) (hc : c.supp ⊆ c'.supp}, c.1.supp = c'.supp := by
```

File: Mathlib/Combinatorics/SimpleGraph/Connectivity/WalkCounting.lean
Comment (YaelDillies) [inline Mathlib/Combinatorics/SimpleGraph/Connectivity/WalkCounting.lean:190]: Do you think you could golf the lemma above using this one? (by setting `G' := ⊤`

Comment (pimotte) [reply Mathlib/Combinatorics/SimpleGraph/Connectivity/WalkCounting.lean:190]: Mainly because I couldn't figure out how to rewrite between these two forms in using it :see_no_evil:

Comment (YaelDillies) [reply Mathlib/Combinatorics/SimpleGraph/Connectivity/WalkCounting.lean:190]: I think you never need your original form. You must have thought you needed it because of your statement of `union_supp_eq_supp`, but my suggestion below should fix this

File: Mathlib/Combinatorics/SimpleGraph/Path.lean
Comment (pimotte) [reply Mathlib/Combinatorics/SimpleGraph/Path.lean]: I tried this, two remarks:

- Is biUnion the right name for this? This is just a double union, whereas the biUnion is over a set, right?
- I'm having trouble using the lemma because the motive is incorrect. Previously I managed to fix this by adding the `DecidablePred`, but I don't really know how to get there now.

File: Mathlib/Combinatorics/SimpleGraph/Connectivity/WalkCounting.lean
Comment (pimotte) [reply Mathlib/Combinatorics/SimpleGraph/Connectivity/WalkCounting.lean]: Done. In number of lines it's not much better, but I'm guessing there's some golfing possibilities

File: Mathlib/Combinatorics/SimpleGraph/Path.lean
Comment (pimotte) [reply Mathlib/Combinatorics/SimpleGraph/Path.lean]: Managed to get this to work, but I'm guessing we want to do some golfing/extraction before merging. In particular the argument to Finset.filter_congr seems a prime candidate. Any thoughts what we would want to do there?

Comment (YaelDillies) [reply Mathlib/Combinatorics/SimpleGraph/Path.lean]: > * Is biUnion the right name for this? This is just a double union, whereas the biUnion is over a set, right?

`biUnion` is precisely the special case of `iUnion₂` where the inner `iUnion` runs over a Prop

File: Mathlib/Combinatorics/SimpleGraph/Connectivity/WalkCounting.lean
Comment (YaelDillies) [inline Mathlib/Combinatorics/SimpleGraph/Connectivity/WalkCounting.lean:212]:
```suggestion
  cases isEmpty_or_nonempty V
```
is more idiomatic

Comment (pimotte) [reply Mathlib/Combinatorics/SimpleGraph/Connectivity/WalkCounting.lean]: Fixed:)

File: Mathlib/Combinatorics/SimpleGraph/Connectivity/WalkCounting.lean
Comment (YaelDillies) [inline Mathlib/Combinatorics/SimpleGraph/Connectivity/WalkCounting.lean:212]: I guess here I meant `cases'` since we don't want to name the hypothesis
```suggestion
  cases' isEmpty_or_nonempty V
```

Comment (pimotte) [reply Mathlib/Combinatorics/SimpleGraph/Connectivity/WalkCounting.lean]: Changed, do I didn't know how to do the second part without naming the hypothesis, so I added a `with` to it

**Comment by pimotte** [comment]
Just backmerged master. Originally I refactored `odd_card_iff_odd_components`, but it had been shown that it could do without [DecidableEq V], and I do think it's needed for the subcomponent variant, so I kept the old stronger result.

**Review by YaelDillies — COMMENTED** [review]
Sorry, I started a review two weeks ago, but it was apparently never published

File: Mathlib/Combinatorics/SimpleGraph/Connectivity/WalkCounting.lean
Comment (pimotte) [reply Mathlib/Combinatorics/SimpleGraph/Connectivity/WalkCounting.lean]: I don't think this is relevant anymore since I reverted back on the old proof to get rid of the dependency on `[DecidableEq V]`, but I'll keep it in mind for future applications

**Review by YaelDillies — APPROVED** [review]
Golfed the proof a bit. Thanks!

maintainer merge

**Review by jcommelin — COMMENTED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**
